Adjusting Expression Evaluator in Calculations for Set Functions

PhilMFischer opened the pull request:
Set-Functions can such as a summary usually summarize a parameter. How deep they should search in the trees can be restricted. So far the implementation was orienting on the levels of nesting regarding the SEIs and their children. Now it is the the level of occurrence of the property or category in question.

Reused the BeanTraverser for this implementation. Generalized it for the usage of SEIs as well.

Change was needed due to the inconsistency between the summary views and the calculations in CEF.

Commit:
Refactored BeanTreeTraverser - (Task #497)

- generalized Traverser to work with any kind of object not just beans.
This is needed as preparation for making it work with pure SEIs as well.
- Added new functionality to know level of nesting and level of matches.
Will be needed for set functions in computations
- Added new test cases
- Added hamcrest lib to test fragment

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

Commit:
Added TreeTraverser to Expression Evaluation - (Task #497)

- Now evaluating Set functions on semantic levels rather than technical
levels.
- Fixed VirSatEcoreUtil to use resolve flag in method as intended
- Added new static variable for infinite nesting

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

Commit:
Added test case for checking traverser in evaluator - (Task #497)

- Added test case to check for correct semantic level detection
- Minor fixes to code to react correctly on comparing matched vs.
required levels of nesting.

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

Commit:
Minor fixes to expression evaluator - (Task #497)

- Fixed issue with cyclic dependencies when e.g. summarizing properties
which are in the same category as the equation including the
set-function.

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

Commit:
Added missing test cases for the Sei Helper - (Task #497)

- A test Case was missing for the get deep children method.

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

Commit:
Adjusted documentation - (Task #497)

- Gave a not ont he emaning of depth.

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

SaMuellerDLR commented on de.dlr.sc.virsat.model.test/src/de/dlr/sc/virsat/model/dvlm/tree/TreeTraverserTest.java:
```diff
@@ -0,0 +1,169 @@
+		public boolean equals(Object obj) {
+			if (this == obj) {
+				return true;
+			} else if (obj == null) {
```
Would it maybe be sufficient to just check 

`if (obj instanceof Result) {
	Result other = (Result) obj;
	return matchingLevel == other.matchingLevel && nestingLevel == other.nestingLevel && Objects.equals(sei, other.sei);
else {
	return false;
}`

That should cover all the cases.

PhilMFischer replied:
Well, the code present is auto coded by the eclipse ide.  The result is basically the same, but there is a difference in performance. Since performance is not an issue in this test case, and in particular not with this object i don't mind changing it.

SaMuellerDLR commented on de.dlr.sc.virsat.model.test/src/de/dlr/sc/virsat/model/dvlm/tree/TreeTraverserTest.java:
```diff
@@ -0,0 +1,169 @@
+			this.sei = sei;
+			this.nestingLevel = processedLevel;
+			this.matchingLevel = matchedLevel;
+		}
```
Can we move these to the top of the `Result` class so the reader directly knows whats encapsualted by it?

PhilMFischer replied:
yes done.

SaMuellerDLR reviewed (changes requested):
Looks good. Only have an issue with a minor readability thing in a test case.

andriikovalov-dlr commented on de.dlr.sc.virsat.model.test/src/de/dlr/sc/virsat/model/dvlm/tree/TreeTraverserTest.java:
```diff
@@ -0,0 +1,169 @@
+			
+			@Override
+			public boolean isMatching(StructuralElementInstance treeNode) {
+				return treeNode.getName() != null;
```
Not immediately clear what it matches. I suggest we either check it explicitly like `return treeNode != seiChild1 && treenode != seiChild2` or check a special name like "MISMATCH", and set it to mismatching elements.

andriikovalov-dlr commented on de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/tree/ITreeTraverserMatcher.java:
```diff
@@ -0,0 +1,57 @@
+	 * @param matchingParent the closest parent of the node that also matches the criteria.
+	 *        Can be null if the node does not have a matching parent
+	 */
+	void processMatch(TYPE treeNode, TYPE matchingParent);
```
Shall we also add tree level and match level as parameters here?

PhilMFischer replied:
I thought about that one as well, but wanted to keep it to as minimal change as needed. thus i would argues for no, for the moment.

andriikovalov-dlr commented on de.dlr.sc.virsat.model.test/src/de/dlr/sc/virsat/model/dvlm/tree/TreeTraverserTest.java:
```diff
@@ -0,0 +1,169 @@
+				
+				// Try to make sure that the seiChild2ChildChild which is theoretically correct, still does not get found
+				boolean isChild2Child = treeNode == seiChild2Child;
+				return !isChild2Child && IStructuralElementInstanceTreeTraverserMatcher.super.continueTraverseChildren(treeNode, isMatching, nestingLevel, matchingLevel);
```
```suggestion
				return treeNode != seiChild2Child;

```

PhilMFischer replied:
The former one, has the default implementation of the interface connected to the test. Thus a change to the default implementation would be captured here. I prefer leaving it as it is, but i added a comment, why the statement has to be that complex.

andriikovalov-dlr reviewed (changes requested):
A couple of minor comments and suggestions

PhilMFischer commented on de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/tree/ITreeTraverserMatcher.java:
```diff
@@ -0,0 +1,57 @@
+	 * @param isMatching the matching result which the traverser evaluated before.
+	 * @return true in case the children should be processed. false in case not.
+	 */
+	default boolean continueTraverseChildren(TYPE treeNode, boolean isMatching, int processedLevel, int matchedLevel) {
```
Good suggestion.

PhilMFischer commented on de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/tree/ITreeTraverserMatcher.java:
```diff
@@ -0,0 +1,57 @@
+	 * @param isMatching the matching result which the traverser evaluated before.
+	 * @return true in case the children should be processed. false in case not.
+	 */
+	default boolean continueTraverseChildren(TYPE treeNode, boolean isMatching, int processedLevel, int matchedLevel) {
```
It is a bit strange checkstyle is not complaining. but probably since the method is that small, it does not expect a javadoc at all.

andriikovalov-dlr replied:
Yes, it's just one line, checkstyle doesn't require a javadoc. But since it's an interface method, it's good to have all parameters explained

PhilMFischer commented on de.dlr.sc.virsat.model.edit/src/de/dlr/sc/virsat/model/concept/types/structural/tree/BeanStructuralTreeTraverser.java:
```diff
@@ -11,55 +11,11 @@
-			}
-		}
-	}
+public class BeanStructuralTreeTraverser extends TreeTraverser<IBeanStructuralElementInstance> {
```
Only for backwards compatibility, because it has been exposed API, and i am not sure about all places where it has been used.

Commit:
Changes due to review comments - (Task #497)

- Adjusted some a java doc.
- made some minor code modifications to improve readability.

---
Task #497: Adjust CEFX calculations to summarize masses and power across
empty system elements

SaMuellerDLR reviewed (approved):
Approved!
